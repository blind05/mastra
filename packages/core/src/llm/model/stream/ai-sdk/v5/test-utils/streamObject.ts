import { delay } from '@ai-sdk/provider-utils';
import { convertAsyncIterableToArray } from '@ai-sdk/provider-utils/test';
import { tool } from 'ai-v5';
import { convertArrayToReadableStream, MockLanguageModelV2, mockValues } from 'ai-v5/test';
import { describe, expect, it } from 'vitest';
import z from 'zod';
import type { execute } from '../../../execute';
import {
  createTestModel,
  defaultSettings,
  modelWithFiles,
  modelWithReasoning,
  modelWithSources,
  testUsage,
  testUsage2,
} from './test-utils';

export function streamObjectTestsV5({ executeFn, runId }: { executeFn: typeof execute; runId: string }) {}
