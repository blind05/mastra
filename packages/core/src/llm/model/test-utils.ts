import type { ServerResponse } from 'http';
import type { LanguageModelV1CallWarning } from '@ai-sdk/provider';
import type { LanguageModelV1, LanguageModelV1StreamPart } from 'ai';
import { MockLanguageModelV1, convertArrayToReadableStream, mockId } from 'ai/test';

const modelWithReasoning = new MockLanguageModelV1({
  doStream: async () => ({
    stream: convertArrayToReadableStream([
      {
        type: 'response-metadata',
        id: 'id-0',
        modelId: 'mock-model-id',
        timestamp: new Date(0),
      },
      { type: 'reasoning', textDelta: 'I will open the conversation' },
      { type: 'reasoning', textDelta: ' with witty banter. ' },
      { type: 'reasoning-signature', signature: '1234567890' },
      { type: 'redacted-reasoning', data: 'redacted-reasoning-data' },
      { type: 'reasoning', textDelta: 'Once the user has relaxed,' },
      {
        type: 'reasoning',
        textDelta: ' I will pry for valuable information.',
      },
      { type: 'reasoning-signature', signature: '1234567890' },
      { type: 'text-delta', textDelta: 'Hi' },
      { type: 'text-delta', textDelta: ' there!' },
      {
        type: 'finish',
        finishReason: 'stop',
        logprobs: undefined,
        usage: { completionTokens: 10, promptTokens: 3 },
      },
    ]),
    rawCall: { rawPrompt: 'prompt', rawSettings: {} },
  }),
});

function createTestModel({
  stream = convertArrayToReadableStream([
    {
      type: 'response-metadata',
      id: 'id-0',
      modelId: 'mock-model-id',
      timestamp: new Date(0),
    },
    { type: 'text-delta', textDelta: 'Hello' },
    { type: 'text-delta', textDelta: ', ' },
    { type: 'text-delta', textDelta: `world!` },
    {
      type: 'finish',
      finishReason: 'stop',
      logprobs: undefined,
      usage: { completionTokens: 10, promptTokens: 3 },
    },
  ]),
  rawCall = { rawPrompt: 'prompt', rawSettings: {} },
  rawResponse = undefined,
  request = undefined,
  warnings,
}: {
  stream?: ReadableStream<LanguageModelV1StreamPart>;
  rawResponse?: { headers: Record<string, string> };
  rawCall?: { rawPrompt: string; rawSettings: Record<string, unknown> };
  request?: { body: string };
  warnings?: LanguageModelV1CallWarning[];
} = {}): LanguageModelV1 {
  return new MockLanguageModelV1({
    doStream: async () => ({ stream, rawCall, rawResponse, request, warnings }),
  });
}

const modelWithSources = new MockLanguageModelV1({
  doStream: async () => ({
    stream: convertArrayToReadableStream([
      {
        type: 'source',
        source: {
          sourceType: 'url' as const,
          id: '123',
          url: 'https://example.com',
          title: 'Example',
          providerMetadata: { provider: { custom: 'value' } },
        },
      },
      { type: 'text-delta', textDelta: 'Hello!' },
      {
        type: 'source',
        source: {
          sourceType: 'url' as const,
          id: '456',
          url: 'https://example.com/2',
          title: 'Example 2',
          providerMetadata: { provider: { custom: 'value2' } },
        },
      },
      {
        type: 'finish',
        finishReason: 'stop',
        logprobs: undefined,
        usage: { completionTokens: 10, promptTokens: 3 },
        providerMetadata: { testprovider: { testkey: 'testvalue' } },
      },
    ]),
    rawCall: { rawPrompt: 'prompt', rawSettings: {} },
  }),
});

const modelWithFiles = new MockLanguageModelV1({
  doStream: async () => ({
    stream: convertArrayToReadableStream([
      {
        type: 'file',
        data: 'Hello World',
        mimeType: 'text/plain',
      },
      { type: 'text-delta', textDelta: 'Hello!' },
      {
        type: 'file',
        data: 'QkFVRw==',
        mimeType: 'image/jpeg',
      },
      {
        type: 'finish',
        finishReason: 'stop',
        logprobs: undefined,
        usage: { completionTokens: 10, promptTokens: 3 },
      },
    ]),
    rawCall: { rawPrompt: 'prompt', rawSettings: {} },
  }),
});

class MockServerResponse {
  writtenChunks: any[] = [];
  headers = {};
  statusCode = 0;
  statusMessage = '';
  ended = false;

  write(chunk: any): void {
    this.writtenChunks.push(chunk);
  }

  end(): void {
    // You might want to mark the response as ended to simulate the real behavior
    this.ended = true;
  }

  writeHead(statusCode: number, statusMessage: string, headers: Record<string, string>): void {
    this.statusCode = statusCode;
    this.statusMessage = statusMessage;
    this.headers = headers;
  }

  get body() {
    // Combine all written chunks into a single string
    return this.writtenChunks.join('');
  }

  /**
   * Get the decoded chunks as strings.
   */
  getDecodedChunks() {
    const decoder = new TextDecoder();
    return this.writtenChunks.map(chunk => decoder.decode(chunk));
  }

  /**
   * Wait for the stream to finish writing to the mock response.
   */
  async waitForEnd() {
    await new Promise(resolve => {
      const checkIfEnded = () => {
        if (this.ended) {
          resolve(undefined);
        } else {
          setImmediate(checkIfEnded);
        }
      };
      checkIfEnded();
    });
  }
}

export function createMockServerResponse(): ServerResponse & MockServerResponse {
  return new MockServerResponse() as ServerResponse & MockServerResponse;
}

export { modelWithReasoning, createTestModel, modelWithSources, modelWithFiles };

export const defaultSettings = () =>
  ({
    prompt: 'prompt',
    experimental_generateMessageId: mockId({ prefix: 'msg' }),
    _internal: {
      generateId: mockId({ prefix: 'id' }),
      currentDate: () => new Date(0),
    },
  }) as const;
