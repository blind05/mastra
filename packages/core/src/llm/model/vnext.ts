import { generateId, type LanguageModel, type ToolSet } from 'ai';
import { z } from 'zod';
import { MessageList } from '../../agent/message-list';
import { MastraAgentStream } from '../../stream/MastraAgentStream';
import { createWorkflow, createStep } from '../../workflows';
import { prepareToolsAndToolChoice } from './prepare-tools';

export class AgenticLoop {
  stepCount = 0;

  toMessageList(messages: any[]) {
    const messageList = new MessageList();

    for (const message of messages) {
      let role;

      if (message.role === 'system') {
        messageList.addSystem(message.content);
        continue;
      }

      if (message.role === 'user') {
        role = 'user';
      } else {
        role = 'response';
      }

      messageList.add(message, role);
    }

    return messageList;
  }

  createToolCallStep({ tools }: { tools?: ToolSet }) {
    return createStep({
      id: 'toolCallStep',
      inputSchema: z.any(),
      outputSchema: z.any(),
      execute: async ({ inputData, getStepResult }) => {
        console.log('Tool call step');

        console.log(JSON.stringify(inputData, null, 2));

        console.log(Object.values(tools || {}));

        const tool =
          tools?.[inputData.toolName] || Object.values(tools || {})?.find(tool => tool.name === inputData.toolName);

        if (!tool) {
          throw new Error(`Tool ${inputData.toolName} not found`);
        }

        if (!tool.execute) {
          return inputData;
        }

        const initialResult = getStepResult({
          id: 'generateText',
        } as any);

        const messageList = this.toMessageList(initialResult.messages);

        const result = await tool.execute(JSON.parse(inputData.args), {
          toolCallId: inputData.toolCallId,
          messages: messageList.get.all.ui().map(message => ({
            role: message.role,
            content: message.content,
          })) as any,
        });

        return { result, ...inputData };
      },
    });
  }

  createLLMStep({
    writer,
    model,
    tools,
    toolChoice,
    providerMetadata,
    runId,
    messageId,
    _internal,
  }: {
    writer: WritableStreamDefaultWriter<any>;
    model: LanguageModel;
    tools?: ToolSet;
    toolChoice?: string;
    providerMetadata?: Record<string, any>;
    runId: string;
    messageId?: string;
    experimental_generateMessageId?: () => string;
    responseId?: string;
    _internal: {
      currentDate: () => Date;
      now: () => number;
      generateId: () => string;
    };
  }) {
    return createStep({
      id: 'generateText',
      inputSchema: z.any(),
      outputSchema: z.any(),
      execute: async ({ inputData }) => {
        await writer.write({
          type: 'step-start',
          payload: {
            request: {},
            warnings: [],
            messageId: messageId,
          },
        });

        this.stepCount++;

        const messageList = this.toMessageList(inputData.messages || []);

        const agentStream = new MastraAgentStream({
          createStream: async () => {
            try {
              const stream = await model.doStream({
                inputFormat: 'messages',
                mode: {
                  type: 'regular',
                  ...prepareToolsAndToolChoice({
                    tools,
                    toolChoice: toolChoice as any,
                    activeTools: Object.keys(tools || {}),
                  }),
                },
                providerMetadata,
                prompt: messageList.get.all.core() as any,
              });

              return stream.stream as any;
            } catch (error) {
              console.error(error);
              return new ReadableStream({
                start: async controller => {
                  controller.enqueue({
                    type: 'error',
                    error,
                  });
                  controller.close();
                },
              });
            }
          },
          getOptions: () => {
            return { runId };
          },
        });

        const defaultResponseMetadata = {
          id: _internal?.generateId?.(),
          timestamp: _internal?.currentDate?.(),
          modelId: model.modelId,
          headers: undefined,
        };

        let responseMetadata: Record<string, any> | undefined = undefined;

        let providerMetadata: Record<string, any> | undefined = undefined;

        let stepFinishPayload;

        let hasErrored = false;

        for await (const chunk of agentStream) {
          switch (chunk.type) {
            case 'error':
              hasErrored = true;
              await writer.write(chunk);
              stepFinishPayload = {
                isContinued: false,
                reason: 'error',
                totalUsage: {
                  promptTokens: 0,
                  completionTokens: 0,
                  totalTokens: 0,
                },
              };
              break;
            case 'response-metadata':
              responseMetadata = {
                id: chunk.payload.id,
                timestamp: chunk.payload.timestamp,
                modelId: chunk.payload.modelId,
                headers: chunk.payload.headers,
              };
              break;
            case 'finish':
              providerMetadata = chunk.payload.providerMetadata;
              stepFinishPayload = {
                reason: chunk.payload.reason,
                logprobs: chunk.payload.logprobs,
                totalUsage: {
                  promptTokens: chunk.payload.totalUsage.promptTokens,
                  completionTokens: chunk.payload.totalUsage.completionTokens,
                  totalTokens:
                    chunk.payload.totalUsage.totalTokens ||
                    chunk.payload.totalUsage.promptTokens + chunk.payload.totalUsage.completionTokens,
                },
                response: responseMetadata || defaultResponseMetadata,
                messageId,
                isContinued: !['stop', 'error'].includes(chunk.payload.reason),
                warnings: chunk.payload.warnings,
                experimental_providerMetadata: chunk.payload.providerMetadata,
                providerMetadata: chunk.payload.providerMetadata,
                request: {},
              };
              break;
            default:
              await writer.write(chunk);
          }

          if (hasErrored) {
            break;
          }
        }

        const text = await agentStream.text;
        const toolCalls = await agentStream.toolCalls;
        const finishReason = await agentStream.finishReason;

        if (toolCalls.length > 0) {
          const userContent = [] as any[];

          if (text) {
            userContent.push({
              type: 'text',
              text: text,
            });
          }

          const assistantContent = [
            ...userContent,
            ...(toolCalls.map(toolCall => {
              return {
                type: 'tool-call',
                toolCallId: toolCall.toolCallId,
                toolName: toolCall.toolName,
                args: JSON.parse(toolCall.args),
              };
            }) as any),
          ];

          messageList.add(
            {
              role: 'assistant',
              content: assistantContent,
            },
            'response',
          );
        } else {
          messageList.add(
            {
              role: 'assistant',
              content: [{ type: 'text', text }],
            },
            'response',
          );
        }

        const messages = messageList.get.all.core();

        const usage = await agentStream.usage;

        console.log('usage', usage);

        return {
          response: {
            stepFinishPayload,
            finishReason: hasErrored ? 'error' : finishReason,
            text,
            toolCalls,
            usage: {
              promptTokens: usage.promptTokens + (inputData.response?.usage?.promptTokens || 0),
              completionTokens: usage.completionTokens + (inputData.response?.usage?.completionTokens || 0),
              totalTokens: usage.promptTokens + usage.completionTokens + (inputData.response?.usage?.totalTokens || 0),
            },
            providerMetadata: providerMetadata,
            metadata: responseMetadata || defaultResponseMetadata,
          },
          messages,
        };
      },
    });
  }

  createExecutionWorkflow({
    writer,
    model,
    tools,
    toolChoice,
    providerMetadata,
    runId,
    experimental_generateMessageId,
    _internal,
  }: {
    writer: WritableStreamDefaultWriter<any>;
    model: LanguageModel;
    tools?: ToolSet;
    toolChoice?: string;
    providerMetadata?: Record<string, any>;
    runId: string;
    responseId?: string;
    experimental_generateMessageId?: () => string;
    _internal: {
      currentDate: () => Date;
      now: () => number;
      generateId: () => string;
    };
  }) {
    const messageId = experimental_generateMessageId?.() || _internal.generateId?.();

    const llmStep = this.createLLMStep({
      writer,
      model,
      tools,
      toolChoice,
      providerMetadata,
      runId,
      messageId,
      _internal,
    });

    const toolCallStep = this.createToolCallStep({ tools });

    return createWorkflow({
      id: 'executionWorkflow',
      inputSchema: z.any(),
      outputSchema: z.any(),
    })
      .then(llmStep)
      .map(({ inputData }) => {
        return inputData.response.toolCalls || [];
      })
      .foreach(toolCallStep)
      .map(async ({ getStepResult, inputData, bail }) => {
        const initialResult = getStepResult(llmStep);

        console.log('stepFinishPayload', initialResult.response.stepFinishPayload);

        if (inputData?.every(toolCall => toolCall?.result === undefined)) {
          await writer.write({
            type: 'step-finish',
            payload: initialResult.response.stepFinishPayload,
          });

          return bail(initialResult);
        }

        const messageList = this.toMessageList(initialResult.messages || []);

        if (inputData?.length) {
          for (const toolCall of inputData) {
            await writer.write({
              type: 'tool-result',
              payload: {
                args: JSON.parse(toolCall.args),
                toolCallId: toolCall.toolCallId,
                toolName: toolCall.toolName,
                result: toolCall.result,
              },
            });
          }

          messageList.add(
            {
              role: 'tool',
              content: inputData.map(toolCall => {
                return {
                  type: 'tool-result',
                  args: JSON.parse(toolCall.args),
                  toolCallId: toolCall.toolCallId,
                  toolName: toolCall.toolName,
                  result: toolCall.result,
                };
              }),
            },
            'response',
          );
        }

        await writer.write({
          type: 'step-finish',
          payload: initialResult.response.stepFinishPayload,
        });

        return {
          ...initialResult,
          messages: messageList.get.all.core(),
        };
      })
      .commit();
  }

  async loop({
    runId,
    model,
    tools,
    toolChoice,
    system,
    prompt,
    threadId,
    resourceId,
    maxSteps = 5,
    maxRetries = 3,
    providerMetadata,
    toolCallStreaming,
    experimental_generateMessageId,
    _internal = {
      currentDate: () => new Date(),
      now: () => Date.now(),
      generateId,
    },
  }: {
    runId?: string;
    model: LanguageModel;
    tools?: ToolSet;
    toolChoice?: string;
    system?: string;
    prompt: string;
    threadId?: string;
    resourceId?: string;
    maxSteps?: number;
    maxRetries?: number;
    toolCallStreaming?: boolean;
    providerMetadata?: Record<string, any>;
    experimental_generateMessageId?: () => string;
    _internal?: {
      currentDate?: () => Date;
      now?: () => number;
      generateId?: () => string;
    };
  }) {
    const messageList = new MessageList({
      threadId,
      resourceId,
    });

    if (system) {
      messageList.addSystem(system);
    }

    if (prompt) {
      messageList.add(prompt, 'user');
    }

    const messages = messageList.get.all.core();

    if (!runId) {
      runId = crypto.randomUUID();
    }

    const agentStream = new MastraAgentStream({
      getOptions: () => {
        return {
          runId: runId,
        };
      },
      createStream: async writer => {
        const readableStream = new ReadableStream({
          start: async controller => {
            const agentStreamWriter = writer.getWriter();

            const workflowStream = new WritableStream({
              write: async chunk => {
                await agentStreamWriter.write(chunk);
              },
            });

            // We call this for no reason because of aisdk
            _internal.generateId?.();

            const workflowStreamWriter = workflowStream.getWriter();

            const executionWorkflow = this.createExecutionWorkflow({
              writer: workflowStreamWriter,
              model,
              tools: tools || undefined,
              toolChoice: toolChoice,
              providerMetadata: providerMetadata,
              runId,
              experimental_generateMessageId,
              _internal,
            });

            const mainWorkflow = createWorkflow({
              id: 'agentic-loop',
              inputSchema: z.any(),
              outputSchema: z.any(),
              retryConfig: {
                attempts: maxRetries,
              },
            })
              .dowhile(executionWorkflow, async ({ inputData }) => {
                return !['stop', 'error'].includes(inputData.response.finishReason) && this.stepCount < maxSteps;
              })
              .map(({ inputData }) => {
                const toolCalls = inputData.messages.filter(message => message.role === 'tool');

                inputData.response.toolCalls = toolCalls;

                return inputData;
              })

              .commit();

            const run = await mainWorkflow.createRunAsync({
              runId,
            });

            const executionResult = await run.start({
              inputData: {
                messages,
              },
            });

            if (executionResult.status !== 'success') {
              controller.close();
              return;
            }

            // console.log('Execution result', executionResult.result.response);

            await workflowStreamWriter.write({
              type: 'finish',
              payload: {
                logprobs: executionResult.result.response.logprobs,
                usage: executionResult.result.response.usage,
                finishReason: executionResult.result.response.finishReason,
                response: executionResult.result?.response?.metadata,
                providerMetadata: executionResult.result?.response?.providerMetadata,
                experimental_providerMetadata: executionResult.result?.response?.providerMetadata,
              },
            });

            controller.close();
          },
        });

        return readableStream as any;
      },
    });

    return agentStream;
  }
}
