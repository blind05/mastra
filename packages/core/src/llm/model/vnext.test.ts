import { convertArrayToReadableStream, convertAsyncIterableToArray } from '@ai-sdk/provider-utils/test';
import { MockLanguageModelV1, mockId } from 'ai/test';
import { describe, expect, it } from 'vitest';

import { createTestModel, modelWithFiles, modelWithReasoning, modelWithSources } from './test-utils';

import { AgenticLoop } from './vnext';

const defaultSettings = () =>
  ({
    prompt: 'prompt',
    experimental_generateMessageId: mockId({ prefix: 'msg' }),
    _internal: {
      generateId: mockId({ prefix: 'id' }),
      currentDate: () => new Date(0),
    },
  }) as const;

const looper = new AgenticLoop();

const runId = '12345';

describe('V4 tests', () => {
  describe('result.textStream', () => {
    it('should send text deltas', async () => {
      const model = new MockLanguageModelV1({
        doStream: async ({ prompt, mode }) => {
          expect(mode).toStrictEqual({
            type: 'regular',
            tools: undefined,
            toolChoice: undefined,
          });

          expect(prompt).toStrictEqual([
            {
              role: 'user',
              content: [{ type: 'text', text: 'test-input' }],
              // providerMetadata: undefined,
            },
          ]);

          return {
            stream: convertArrayToReadableStream([
              { type: 'text-delta', textDelta: 'Hello' },
              { type: 'text-delta', textDelta: ', ' },
              { type: 'text-delta', textDelta: `world!` },
              {
                type: 'finish',
                finishReason: 'stop',
                logprobs: undefined,
                usage: { completionTokens: 10, promptTokens: 3 },
              },
            ]),
            rawCall: { rawPrompt: 'prompt', rawSettings: {} },
          };
        },
      });

      const result = await looper.loop({
        runId,
        model,
        system: 'You are a helpful assistant.',
        prompt: 'test-input',
        threadId: '123',
        resourceId: '456',
      });

      expect(await convertAsyncIterableToArray(result.textStream)).toStrictEqual(['Hello', ', ', 'world!']);
    });

    it('should filter out empty text deltas', async () => {
      const result = await looper.loop({
        runId,
        model: createTestModel({
          stream: convertArrayToReadableStream([
            { type: 'text-delta', textDelta: '' },
            { type: 'text-delta', textDelta: 'Hello' },
            { type: 'text-delta', textDelta: '' },
            { type: 'text-delta', textDelta: ', ' },
            { type: 'text-delta', textDelta: '' },
            { type: 'text-delta', textDelta: 'world!' },
            { type: 'text-delta', textDelta: '' },
            {
              type: 'finish',
              finishReason: 'stop',
              logprobs: undefined,
              usage: { completionTokens: 10, promptTokens: 3 },
            },
          ]),
        }),
        system: 'You are a helpful assistant.',
        prompt: 'test-input',
        threadId: '123',
        resourceId: '456',
      });

      expect(await convertAsyncIterableToArray(result.textStream)).toMatchSnapshot();
    });

    it('should not include reasoning content in textStream', async () => {
      const result = await looper.loop({
        model: modelWithReasoning,
        runId,
        ...defaultSettings(),
      });

      expect(await convertAsyncIterableToArray(result.textStream)).toMatchSnapshot();
    });

    it('should swallow error to prevent server crash', async () => {
      const result = await looper.loop({
        runId,
        model: new MockLanguageModelV1({
          doStream: async () => {
            throw new Error('test error');
          },
        }),
        prompt: 'test-input',
      });

      expect(await convertAsyncIterableToArray(result.textStream)).toMatchSnapshot();
    });
  });

  describe('result.fullStream', () => {
    it('should send text deltas', async () => {
      const result = await looper.loop({
        runId,
        model: new MockLanguageModelV1({
          doStream: async ({ prompt, mode }) => {
            expect(mode).toStrictEqual({
              type: 'regular',
              tools: undefined,
              toolChoice: undefined,
            });

            expect(prompt).toStrictEqual([
              {
                role: 'user',
                content: [{ type: 'text', text: 'test-input' }],
                // providerMetadata: undefined,
              },
            ]);

            return {
              stream: convertArrayToReadableStream([
                {
                  type: 'response-metadata',
                  id: 'response-id',
                  modelId: 'response-model-id',
                  timestamp: new Date(5000),
                },
                { type: 'text-delta', textDelta: 'Hello' },
                { type: 'text-delta', textDelta: ', ' },
                { type: 'text-delta', textDelta: `world!` },
                {
                  type: 'finish',
                  finishReason: 'stop',
                  logprobs: undefined,
                  usage: { completionTokens: 10, promptTokens: 3 },
                },
              ]),
              rawCall: { rawPrompt: 'prompt', rawSettings: {} },
            };
          },
        }),
        prompt: 'test-input',
        experimental_generateMessageId: mockId({ prefix: 'msg' }),
      });

      const messages = await convertAsyncIterableToArray(result.aisdkv4);

      expect(messages).toMatchSnapshot();
    });

    it('should send reasoning deltas', async () => {
      const result = await looper.loop({
        runId,
        model: modelWithReasoning,
        system: 'You are a helpful assistant.',
        prompt: 'test-input',
        threadId: '123',
        resourceId: '456',
        experimental_generateMessageId: mockId({ prefix: 'msg' }),
      });

      expect(await convertAsyncIterableToArray(result.aisdkv4)).toMatchSnapshot();
    });

    it('should send sources', async () => {
      const result = await looper.loop({
        runId,
        model: modelWithSources,
        system: 'You are a helpful assistant.',
        threadId: '123',
        resourceId: '456',
        ...defaultSettings(),
      });

      expect(await convertAsyncIterableToArray(result.aisdkv4)).toMatchSnapshot();
    });

    it.only('should send files', async () => {
      const result = await looper.loop({
        model: modelWithFiles,
        ...defaultSettings(),
      });

      expect(await convertAsyncIterableToArray(result.aisdkv4)).toMatchSnapshot();
    });
  });

  it('should send files', async () => {
    const result = await looper.loop({
      runId,
      model: modelWithFiles,
      system: 'You are a helpful assistant.',
      prompt: 'test-input',
      threadId: '123',
      resourceId: '456',
    });

    expect(await convertAsyncIterableToArray(result)).toMatchSnapshot();
  });

  it('should use fallback response metadata when response metadata is not provided', async () => {
    const result = await looper.loop({
      runId,
      model: new MockLanguageModelV1({
        doStream: async ({ prompt, mode }) => {
          expect(mode).toStrictEqual({
            type: 'regular',
            tools: undefined,
            toolChoice: undefined,
          });

          expect(prompt).toStrictEqual([
            {
              role: 'user',
              content: [{ type: 'text', text: 'test-input' }],
              providerMetadata: undefined,
            },
          ]);

          return {
            stream: convertArrayToReadableStream([
              { type: 'text-delta', textDelta: 'Hello' },
              { type: 'text-delta', textDelta: ', ' },
              { type: 'text-delta', textDelta: `world!` },
              {
                type: 'finish',
                finishReason: 'stop',
                logprobs: undefined,
                usage: { completionTokens: 10, promptTokens: 3 },
              },
            ]),
            rawCall: { rawPrompt: 'prompt', rawSettings: {} },
          };
        },
      }),
      prompt: 'test-input',
      // _internal: {
      //     currentDate: mockValues(new Date(2000)),
      //     generateId: mockValues('id-2000'),
      // },
      // experimental_generateMessageId: mockId({ prefix: 'msg' }),
    });

    console.log(await convertAsyncIterableToArray(result));

    expect(await convertAsyncIterableToArray(result)).toMatchSnapshot();
  });

  describe('result.finishReason', () => {
    it('should resolve with finish reason', async () => {
      const result = await looper.loop({
        runId,
        model: createTestModel({
          stream: convertArrayToReadableStream([
            { type: 'text-delta', textDelta: 'Hello' },
            {
              type: 'finish',
              finishReason: 'stop',
              logprobs: undefined,
              usage: { completionTokens: 10, promptTokens: 3 },
            },
          ]),
        }),
        prompt: 'test-input',
      });

      // result.consumeStream();

      expect(await result.finishReason).toStrictEqual('stop');
    });
  });
});
