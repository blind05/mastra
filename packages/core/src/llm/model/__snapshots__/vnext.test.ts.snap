// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`V4 tests > result.fullStream > should send files 1`] = `
[
  {
    "messageId": "msg-0",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  DefaultGeneratedFileWithType {
    "base64Data": "Hello World",
    "mimeType": "text/plain",
    "type": "file",
    "uint8ArrayData": undefined,
  },
  {
    "textDelta": "Hello!",
    "type": "text-delta",
  },
  DefaultGeneratedFileWithType {
    "base64Data": "QkFVRw==",
    "mimeType": "image/jpeg",
    "type": "file",
    "uint8ArrayData": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-0",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-1",
      "modelId": "mock-model-id",
      "timestamp": 1970-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 3,
      "totalTokens": 13,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "id-1",
      "modelId": "mock-model-id",
      "timestamp": 1970-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 3,
      "totalTokens": 13,
    },
  },
]
`;

exports[`V4 tests > result.fullStream > should send reasoning deltas 1`] = `
[
  {
    "messageId": "msg-0",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "textDelta": "I will open the conversation",
    "type": "reasoning",
  },
  {
    "textDelta": " with witty banter. ",
    "type": "reasoning",
  },
  {
    "signature": "1234567890",
    "type": "reasoning-signature",
  },
  {
    "data": "redacted-reasoning-data",
    "type": "redacted-reasoning",
  },
  {
    "textDelta": "Once the user has relaxed,",
    "type": "reasoning",
  },
  {
    "textDelta": " I will pry for valuable information.",
    "type": "reasoning",
  },
  {
    "signature": "1234567890",
    "type": "reasoning-signature",
  },
  {
    "textDelta": "Hi",
    "type": "text-delta",
  },
  {
    "textDelta": " there!",
    "type": "text-delta",
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-0",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-0",
      "modelId": "mock-model-id",
      "timestamp": 1970-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 3,
      "totalTokens": 13,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "id-0",
      "modelId": "mock-model-id",
      "timestamp": 1970-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 3,
      "totalTokens": 13,
    },
  },
]
`;

exports[`V4 tests > result.fullStream > should send sources 1`] = `
[
  {
    "messageId": "msg-0",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "source": {
      "id": "123",
      "providerMetadata": {
        "provider": {
          "custom": "value",
        },
      },
      "sourceType": "url",
      "title": "Example",
      "url": "https://example.com",
    },
    "type": "source",
  },
  {
    "textDelta": "Hello!",
    "type": "text-delta",
  },
  {
    "source": {
      "id": "456",
      "providerMetadata": {
        "provider": {
          "custom": "value2",
        },
      },
      "sourceType": "url",
      "title": "Example 2",
      "url": "https://example.com/2",
    },
    "type": "source",
  },
  {
    "experimental_providerMetadata": {
      "testprovider": {
        "testkey": "testvalue",
      },
    },
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-0",
    "providerMetadata": {
      "testprovider": {
        "testkey": "testvalue",
      },
    },
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-1",
      "modelId": "mock-model-id",
      "timestamp": 1970-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 3,
      "totalTokens": 13,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": {
      "testprovider": {
        "testkey": "testvalue",
      },
    },
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": {
      "testprovider": {
        "testkey": "testvalue",
      },
    },
    "response": {
      "headers": undefined,
      "id": "id-1",
      "modelId": "mock-model-id",
      "timestamp": 1970-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 3,
      "totalTokens": 13,
    },
  },
]
`;

exports[`V4 tests > result.fullStream > should send text deltas 1`] = `
[
  {
    "messageId": "msg-0",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "textDelta": "Hello",
    "type": "text-delta",
  },
  {
    "textDelta": ", ",
    "type": "text-delta",
  },
  {
    "textDelta": "world!",
    "type": "text-delta",
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-0",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "response-id",
      "modelId": "response-model-id",
      "timestamp": 1970-01-01T00:00:05.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 3,
      "totalTokens": 13,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "response-id",
      "modelId": "response-model-id",
      "timestamp": 1970-01-01T00:00:05.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 3,
      "totalTokens": 13,
    },
  },
]
`;

exports[`V4 tests > result.textStream > should filter out empty text deltas 1`] = `
[
  "Hello",
  ", ",
  "world!",
]
`;

exports[`V4 tests > result.textStream > should not include reasoning content in textStream 1`] = `
[
  "Hi",
  " there!",
]
`;

exports[`V4 tests > result.textStream > should swallow error to prevent server crash 1`] = `[]`;
